### ISSUE TITLE
Endless cycle

### ISSUE THREAD

**Issue by ikrivosheev** [mdsteele/rust-cfb#7]
Hello. I have a stream of different files (doc, msi, cfb,cab, ...). I am detecting the file type and sometimes I can get the falsepositive result (for example I got **.doc** but detect as **.msi**) . Then I send the file to unpack this library and panic happens in the debug build. 

Can I make PR where change API, remove `debug_assert` and change to return `Result`?

**Comment by mdsteele** [comment]
Agreed that `rust-cfb` should not panic when trying to read or operate on a malformed file; it should just return an error `Result`.  Right now it tries to validate well-formed-ness when it first opens the file (and return an error `Result` immediately if validation fails), but probably the validation logic is still missing some things.

Can you give an example of the kind of `debug_assert` failures you're seeing?

**Comment by ikrivosheev** [comment]
```
Breakpoint 1, <engine::unpacker::MSIUnpacker as engine::unpacker::Unpacker>::unpack (self=0x7fffffff6df8, file=..., max_file_size=104857600) at engine/src/unpacker.rs:32
32              let archive = msi::Package::open(file)?;
(gdb) n
thread 'main' panicked at 'assertion failed: sector_id < self.num_sectors', /home/ikrivosheev/.cargo/registry/src/github.com-1ecc6299db9ec823/cfb-0.4.0/src/internal/sector.rs:65:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:515:5
   1: core::panicking::panic_fmt
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/panicking.rs:92:14
   2: core::panicking::panic
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/panicking.rs:50:5
   3: cfb::internal::sector::Sectors<F>::seek_within_sector
             at /home/ikrivosheev/.cargo/registry/src/github.com-1ecc6299db9ec823/cfb-0.4.0/src/internal/sector.rs:65:9
   4: cfb::internal::sector::Sectors<F>::seek_to_sector
             at /home/ikrivosheev/.cargo/registry/src/github.com-1ecc6299db9ec823/cfb-0.4.0/src/internal/sector.rs:57:9
   5: cfb::CompoundFile<F>::open
             at /home/ikrivosheev/.cargo/registry/src/github.com-1ecc6299db9ec823/cfb-0.4.0/src/lib.rs:605:30
   6: msi::internal::package::Package<F>::open
             at /home/ikrivosheev/.cargo/registry/src/github.com-1ecc6299db9ec823/msi-0.3.0/src/internal/package.rs:264:24
   7: <engine::unpacker::MSIUnpacker as engine::unpacker::Unpacker>::unpack
             at ./engine/src/unpacker.rs:32:23
   8: engine::engine::Engine::scan_archive
             at ./engine/src/engine.rs:202:15
   9: engine::engine::Engine::scan_file
             at ./engine/src/engine.rs:232:25
  10: scanner::main
             at ./engine/src/main.rs:88:32
  11: core::ops::function::FnOnce::call_once
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by mdsteele** [comment]
Thanks.  I've just pushed commit 9ed6ee0365811798125c6d979d813db68a14782d to turn that particular `debug_assert` into a `Result` error, and also do better up-front validation of the FAT and DIFAT chains to catch the problem earlier (without panicking).

Let me know if that helps, and if you run into any new panics.

**Comment by ikrivosheev** [comment]
Thank you for fix! Can you make release and update dependencies for [rust-msi](https://github.com/mdsteele/rust-msi)?

**Comment by ikrivosheev** [comment]
@mdsteele, cool! It works!

**Comment by mdsteele** [comment]
Sure.  I've just published [`cfb` v0.5.0](https://crates.io/crates/cfb/0.5.0), and also [`msi` v0.3.1](https://crates.io/crates/msi/0.3.1) which uses it.
